### ISSUE TITLE
Bad type inference when calling the same generic function twice in one statement

### ISSUE THREAD

**Issue by BeatButton** [microsoft/pyright#7245]
When calling a function that returns a `TypeVar` multiple times in the same statement, pyright incorrectly infers that the second call must have the `TypeVar` bound the same way.

```python
from typing import TypeVar

T = TypeVar("T")
def foo(v: T) -> T:
    return v

if (x := foo(1)) and (y := foo("s")):
    print(x, y)

```
pyright output:
```
error: Argument of type "Literal['s']" cannot be assigned to parameter "v" of type "T@foo" in function "foo"
    "Literal['s']" is incompatible with "int" (reportArgumentType)
```

Error produced using:
Pylance VSC extension, v2024.2.1
pyright command line 1.1.350

**Comment by erictraut** [comment]
This isn't related to calling the same function twice. The same behavior occurs in this example:

```python
from typing import TypeVar

T = TypeVar("T")

def foo(v: T) -> T:
    return v

a = 1 and foo("s")
```

The reason you're seeing this behavior is that pyright is using bidirectional type inference when inferring the type of the second operand of `and` and `or` operators. This support was added for situations like:

```python
def func(x: list[str] | None):
    y = x and []
```
In this case, you want the second operand (which would normally have an inferred type of `list[Unknown]`) to be inferred as `list[str]`, consistent with the first operand.

I'll need to see if I can find a way to make your example work without adversely affecting this other use case.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.352, which I just published.
